rfcs/2023-01-06-layout.md

```diff
@@ -0,0 +1,460 @@
+
+A brief study of some layout primitives is compiled in the [references](#references) section below.
+
+### Layout primitives and components for Blade
+
+In order to provide a good balance of flexibility and constraints within blade we should support the theme aware components and primitives listed below.
+
+#### `Box` component
+
+A general purpose component for building layouts. This component will do a bulk of heavy lifting for composing layouts. Sample usage:
+
+```tsx
+<Box
+  display="flex"
+  flexDirection="column"
+  alignItems="center"
+  marginTop="spacing.1"
+  marginBottom="spacing.2"
+  paddingX="spacing.1"
+  gap="spacing.1"
+>
+  <Button>Button 1</Button>
+  <Button>Button 2</Button>
+  <Button>Button 3</Button>
+</Box>
+```
+
+Since blade is not a general purpose design system we should lean more towards constraints initially, assess usecases in the future and then increase flexibility. Proposed props for initial version:
+
+| Category            | Props                                       | Notes                                                                                                                                  |
+| ------------------- | ------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------- |
+| Spacing             | `margin*`, `padding*`,                      | [List of props](https://github.com/styled-system/styled-system/blob/master/docs/table.md#space) referenced from styled-system.         |
+| Layout              | `width*`, `height*`, `overflow*`, `display` | [List of props](https://github.com/styled-system/styled-system/blob/master/docs/table.md#layout) referenced from styled-system.        |
+| Flexbox             | `flex*`, `align*`, `justify*`, `order`      | [List of props](https://github.com/styled-system/styled-system/blob/master/docs/table.md#flexbox) referenced from styled-system.       |
+| Gap                 | `gap`, `rowGap`, `columnGap`                | Provides native flex functionality for spacing siblings.                                                                               |
+| Position            | `position`, `zIndex`, `top`, etc.           | [List of props](https://github.com/styled-system/styled-system/blob/master/docs/table.md#position) referenced from styled-system.      |
+| Grid **(web only)** | `grid*`                                     | [List of props](https://github.com/styled-system/styled-system/blob/master/docs/table.md#grid-layout)\* referenced from styled-system. |
+
+> **Note**
+>
+> Not all CSS grid properties are covered in `styled-system` spec. These should be referenced from MDN during implementation. Here's a nice reference from [CSS Tricks](https://css-tricks.com/snippets/css/complete-guide-grid).
```

Comment:
The CSS grid properties are very extensive and for a reason `styled-system` supports only a few of them. We should limit the scope for the number of properties we should include otherwise the execution can be painstaking. I would vote for keeping it synchronous with what `styled-system` exposes.

Reply:
I found some useful properties (eg. `grid-row-start`, `grid-column-start`, etc.) weren't documented in the linked styled-system doc so added this as reference for confirming during implementation.

Implementation wise we would be mostly just exposing these as props and passing them as CSS properties, so I think it shouldn't affect execution.